File: built-in plugins dsp/mb-pad-digital.cpp
/**
 * MB Digital Pad
 * Category : instrument
 * Type     : pad
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Clean digital pad
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_PAD_DIGITAL_H
#define MB_PAD_DIGITAL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbPadDigital : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-pad-digital";
    static constexpr const char* PLUGIN_NAME    = "MB Digital Pad";
    static constexpr const char* PLUGIN_TYPE    = "pad";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float shimmer = 0.5f;  // range [0, 1]
    float volume = 0.7f;  // range [0, 1]
    };

    MbPadDigital() = default;
    ~MbPadDigital() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.shimmer = std::clamp(params.shimmer, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Digital Pad
        return input;
    }
};

#endif // MB_PAD_DIGITAL_H

File: built-in plugins dsp/mb-pad-digital.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbPadDigitalPlugin: PluginDefinition = {
    id: 'mb-pad-digital', slug: 'mb-pad-digital', name: 'MB Digital Pad', category: 'instrument', type: 'pad', version: '1.0.0',
    description: 'Clean digital pad', author: 'Max Booster',
    oscillators: [{ type: 'sine', detune: 0, gain: 0.4 }, { type: 'triangle', detune: 1200, gain: 0.3 }, { type: 'sine', detune: 1900, gain: 0.3 }],
    envelope: { attack: 0.5, decay: 0.8, sustain: 0.85, release: 1.5 },
    parameters: [
      { id: 'shimmer', name: 'Shimmer', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { shimmer: 0.5, volume: 0.7 },
  };

export default MbPadDigitalPlugin;
